3.16.0 regression: padded-blocks autofix can generate incorrect code

**Tell us about your environment**

* **ESLint Version:** 3.16.0
* **Node Version:** 6.9.1
* **npm Version:** 3.10.8

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

```
module.exports = {
  rules: {
    "padded-blocks": ['error', 'never'],
  }
};
```

(This is the value specified by eslint-config-airbnb.)

**What did you do? Please include the actual source code causing the issue.**

I ran `eslint --fix` on this source code:

```js
function f() { // a

  b;
}
```

**What did you expect to happen?**

I expected it to fix the formatting and generate correct code.

**What actually happened? Please include the actual, raw output from ESLint.**

eslint generated this output:
```
➜  eslint_test_3 git:(master) ./node_modules/.bin/eslint --fix test.js

/Users/alanpierce/code/eslint_test_3/test.js
  6:1  error  Parsing error: Unexpected token

✖ 1 problem (1 error, 0 warnings)
```

And the JS file was changed to have these contents:
```
function f() {
function f() { // a

  b;
}
```

This is a syntax error, so the file fails to parse after this change.

When I try this on eslint 3.15.0, it doesn't have any problems.

This is breaking the [coffeelint example build](https://github.com/decaffeinate/decaffeinate#status) for the decaffeinate project. The [lexical_linter.js](https://github.com/decaffeinate-examples/coffeelint/blob/decaffeinate/src/lexical_linter.js) file has two `static initClass` lines near the start, which causes a syntax error. (The example builds are a bit like eslint-canary for the decaffeinate project, and one of the conversion steps is to run `eslint --fix` on the resulting JS code.)

--- a/lib/rules/padded-blocks.js
+++ b/lib/rules/padded-blocks.js
@@ -188,6 +188,9 @@ module.exports = {
                         node,
                         loc: { line: openBrace.loc.start.line, column: openBrace.loc.start.column },
                         fix(fixer) {
+
+                            // FIXME: The start of this range is sometimes larger than the end.
+                            // https://github.com/eslint/eslint/issues/8116
                             return fixer.replaceTextRange([openBrace.end, nextToken.start - nextToken.loc.start.column], "\n");
                         },
                         message: NEVER_MESSAGE

--- a/lib/util/source-code-fixer.js
+++ b/lib/util/source-code-fixer.js
@@ -108,7 +108,13 @@ SourceCodeFixer.applyFixes = function(sourceCode, messages) {
             // Make output to this fix.
             output += text.slice(Math.max(0, lastPos), Math.max(0, start));
             output += fix.text;
-            lastPos = end;
+
+            /*
+             * If the start of the range is larger than the end for some reason, make sure
+             * the text between the end and the start doesn't get duplicated.
+             * https://github.com/eslint/eslint/issues/8116
+             */
+            lastPos = Math.max(start, end);
         }
         output += text.slice(Math.max(0, lastPos));
 

--- a/tests/lib/rules/padded-blocks.js
+++ b/tests/lib/rules/padded-blocks.js
@@ -463,6 +463,12 @@ ruleTester.run("padded-blocks", rule, {
                     line: 9
                 }
             ]
+        },
+        {
+            code: "function foo() { // a\n\n  b;\n}",
+            output: "function foo() {\n // a\n\n  b;\n}",
+            options: ["never"],
+            errors: [NEVER_MESSAGE]
         }
     ]
 });

--- a/tests/lib/util/source-code-fixer.js
+++ b/tests/lib/util/source-code-fixer.js
@@ -122,6 +122,13 @@ const INSERT_AT_END = {
         message: "nofix2",
         line: 1,
         column: 7
+    },
+    REVERSED_RANGE = {
+        message: "reversed range",
+        fix: {
+            range: [3, 0],
+            text: " "
+        }
     };
 
 //------------------------------------------------------------------------------
@@ -405,6 +412,12 @@ describe("SourceCodeFixer", () => {
 
                 assert.equal(result.output, `\uFEFF${INSERT_AT_START.fix.text}${insertInMiddle}${INSERT_AT_END.fix.text}`);
                 assert.equal(result.messages.length, 0);
+            });
+
+            it("should handle reversed ranges gracefully", () => {
+                const result = SourceCodeFixer.applyFixes(sourceCode, [REVERSED_RANGE]);
+
+                assert.equal(result.output, "\uFEFFvar  answer = 6 * 7;");
             });
 
         });

